### ISSUE TITLE
[http] Binding does not handle 401 protocol violations gracefully 

### ISSUE THREAD

**Issue by mrjvb** [openhab/openhab-addons#16110]
## Expected Behavior

HTTP GET request using digest authentication to (for example) a doorbell camera to the following address:

`http://192.168.1.XX/ISAPI/VideoIntercom/callStatus`

should negotiate authentication and return the answer on first **and subsequent** calls.

## Current Behavior

The call succeeds **only once**. When the camera returns a 401 the binding does not attempt to obtain a new digest authentication and keeps trying with the rejected authorization values.

## Possible Solution

Binding should (probably) re-negotiate the digest authentication upon receiving a 401 instead of continuing to use the cached credentials.

## Steps to Reproduce (for Bugs)

1. HTTP URL thing defined as:
```
UID: http:url:doorbellpoll
label: Doorbell Poll
thingTypeUID: http:url
configuration:
  authMode: DIGEST
  ignoreSSLErrors: false
  baseURL: http://192.168.1.XX/ISAPI/VideoIntercom/callStatus
  password: password
  delay: 0
  stateMethod: GET
  refresh: 30
  commandMethod: GET
  timeout: 3000
  username: user
  bufferSize: 2048
```
2. First TRACE log:
```
[TRACE] [.HttpDynamicStateDescriptionProvider] - removing state description for thing http:url:doorbellpoll
[INFO ] [nding.http.internal.HttpThingHandler] - Using the secure client for thing 'http:url:doorbellpoll'.
[DEBUG] [nding.http.internal.HttpThingHandler] - Digest Authentication configured for thing 'http:url:doorbellpoll'
[TRACE] [ttp.internal.http.RefreshingUrlCache] - Started refresh task for URL 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus' with interval 30s
[TRACE] [.HttpDynamicStateDescriptionProvider] - adding state description for channel http:url:doorbellpoll:doorbellpollstatus
[TRACE] [ttp.internal.http.RefreshingUrlCache] - Stopped refresh task for URL 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus'
[TRACE] [.HttpDynamicStateDescriptionProvider] - removing state description for thing http:url:doorbellpoll
[INFO ] [nding.http.internal.HttpThingHandler] - Using the secure client for thing 'http:url:doorbellpoll'.
[DEBUG] [nding.http.internal.HttpThingHandler] - Digest Authentication configured for thing 'http:url:doorbellpoll'
[TRACE] [ttp.internal.http.RefreshingUrlCache] - Started refresh task for URL 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus' with interval 30s
[TRACE] [.HttpDynamicStateDescriptionProvider] - adding state description for channel http:url:doorbellpoll:doorbellpollstatus
[TRACE] [ttp.internal.http.RefreshingUrlCache] - Requesting refresh (retry=false) from 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus' with timeout 3000ms
[TRACE] [ttp.internal.http.RefreshingUrlCache] - Sending to 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus': Method = {GET}, Headers = {Accept-Encoding: gzip, User-Agent: Jetty/9.4.52.v20230823}, Content = {null}
[TRACE] [p.internal.http.HttpResponseListener] - Received from 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus': Code = {200}, Headers = {Date: Sun, 24 Dec 2023 10:49:13 GMT, Server: webs, Content-Length: 42, Connection: close, X-Frame-Options: SAMEORIGIN, Cache-Control: no-store, Pragma: no-cache, Content-Type: application/xml}, Content = {{
	"CallStatus":	{
		"status":	"idle"
	}
}}
```
3. Next TRACE log:
```
[TRACE] [ttp.internal.http.RefreshingUrlCache] - Requesting refresh (retry=false) from 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus' with timeout 3000ms
[TRACE] [ttp.internal.http.RefreshingUrlCache] - Sending to 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus': Method = {GET}, Headers = {Accept-Encoding: gzip, User-Agent: Jetty/9.4.52.v20230823}, Content = {null}
[TRACE] [p.internal.http.HttpResponseListener] - Received from 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus': Code = {401}, Headers = {Date: Sun, 24 Dec 2023 10:49:42 GMT, Server: webs, Content-Length: 387, Connection: close, X-Frame-Options: SAMEORIGIN, Cache-Control: no-store, Pragma: no-cache, Content-Type: application/xml}, Content = {<?xml version="1.0" encoding="UTF-8"?>
<ResponseStatus version="1.0" xmlns="http://www.std-cgi.com/ver10/XMLSchema">
<requestURL>/ISAPI/VideoIntercom/callStatus</requestURL>
<statusCode>4</statusCode>
<statusString>Invalid Operation</statusString>
<subStatusCode>invalidOperation</subStatusCode>
<errorCode>1073741830</errorCode>
<errorMsg>invalid operation</errorMsg>
</ResponseStatus>
}

[WARN ] [p.internal.http.HttpResponseListener] - Requesting 'http://192.168.1.XX/ISAPI/VideoIntercom/callStatus' (method='GET', content='null') failed: org.eclipse.jetty.client.HttpResponseException: HTTP protocol violation: Authentication challenge without WWW-Authenticate header
```
4. Package capture:

```
* First attempt works *
GET /ISAPI/VideoIntercom/callStatus HTTP/1.1
Accept-Encoding: gzip
User-Agent: Jetty/9.4.52.v20230823
Authorization: Digest username="user", realm="DS-1762ACD3", nonce="NTU2MDI3MTlmNGIwZmNkODZmMzg4OTdiZWI2ZGZhYzk=", opaque="", algorithm="MD5", uri="/ISAPI/VideoIntercom/callStatus", qop="auth", nc="00000001", cnonce="4a67b856d0345194", response="99f08809ce0811b236f793eccecc9bdb"
Host: 192.168.1.XX

* This response *
HTTP/1.1 200 OK
Date: Sun, 24 Dec 2023 13:09:07 GMT
Server: webs
Content-Length: 42
Connection: close
X-Frame-Options: SAMEORIGIN
Cache-Control: no-store
Pragma: no-cache
Content-Type: application/xml

{
	"CallStatus":	{
		"status":	"idle"
	}
}

* Next attempt fails *
GET /ISAPI/VideoIntercom/callStatus HTTP/1.1
Accept-Encoding: gzip
User-Agent: Jetty/9.4.52.v20230823
Host: 192.168.1.XX
Authorization: Digest username="user", realm="DS-1762ACD3", nonce="NTU2MDI3MTlmNGIwZmNkODZmMzg4OTdiZWI2ZGZhYzk=", opaque="", algorithm="MD5", uri="/ISAPI/VideoIntercom/callStatus", qop="auth", nc="00000002", cnonce="81a72588115fc84c", response="18adfcd5dc6d0427ce31da7210a82327"

HTTP/1.1 401 Unauthorized
Date: Sun, 24 Dec 2023 13:09:19 GMT
Server: webs
Content-Length: 387
Connection: close
X-Frame-Options: SAMEORIGIN
Cache-Control: no-store
Pragma: no-cache
Content-Type: application/xml

<?xml version="1.0" encoding="UTF-8"?>
<ResponseStatus version="1.0" xmlns="http://www.std-cgi.com/ver10/XMLSchema">
<requestURL>/ISAPI/VideoIntercom/callStatus</requestURL>
<statusCode>4</statusCode>
<statusString>Invalid Operation</statusString>
<subStatusCode>invalidOperation</subStatusCode>
<errorCode>1073741830</errorCode>
<errorMsg>invalid operation</errorMsg>
</ResponseStatus>
```

## Your Environment

* Version used: openHAB 4.1.0 and org.openhab.binding.http 4.1.0
* OS: Debian 11.8

**Comment by J-N-K** [comment]
That's very interesting. According to the code You should run into one of these messages. But I don't see them in your log. Can you confirm that they are really not logged if you set the binding to TRACE?

```
logger.warn("Retry after authentication failure failed again for                                 logger.debug("Cleared authentication result for '{}', retrying immediately", uri);
logger.warn("Could not find authentication result for '{}', failing here", uri);
```

**Comment by J-N-K** [comment]
Looking at it a bit more: The response from the server is invalid on the second request. If a server responds with 401/Unauthorized it has to include a `WWW-Authenticate` header in the response [RFC-7235, Section 4.1](https://datatracker.ietf.org/doc/html/rfc7235#section-4.1):

> A server generating a 401 (Unauthorized) response MUST send a WWW-Authenticate header field containing at least one challenge.  A server MAY generate a WWW-Authenticate header field in other response messages to indicate that supplying credentials (or different credentials) might affect the response.

Since curl (and probably others) don't cache the authentication result, they always re-auth. Jetty does cache them (and that works fine if the server follows the standard) and runs into an invalid 401 response. 

A new authentication method "One-time Digest" that clears the result after each request would be a work-around to this. The other solution would be to remove the `WWWAuthenticationProtocolHandler` from the client and do the authentication flow in our own code.

**Comment by mrjvb** [comment]
Thank you very much for looking into this. No messages like these:

> That's very interesting. According to the code You should run into one of these messages. But I don't see them in your log. Can you confirm that they are really not logged if you set the binding to TRACE?
> 
> ```
> logger.warn("Retry after authentication failure failed again for                                 logger.debug("Cleared authentication result for '{}', retrying immediately", uri);
> logger.warn("Could not find authentication result for '{}', failing here", uri);
> ```
appear in the log with the binding on TRACE logging. Your conclusion that the response from the server (Hikvision camera in this case) with a 401 without WWW-Authenticate is non-compliant to RFC-7235 is valid. A "One-time Digest" option would be great. Not sure if there is a "catch-all" way, for example to clear cached credentials once a 401 response is received without the WWW-Authenticate header field?

**Comment by J-N-K** [comment]
The credentials are cleared when a 401 is received. But in this case Jetty already complains about an invalid response before that code is even reached.

**Comment by mrjvb** [comment]
Understood, the server (camera) is not following the RFC-7235, and the binding relies on Jetty, so not a bug (in either OH or Jetty). Not sure if my use-case is unique, but if a fix is possible to prevent the binding from endlessly retrying nc="00000002", nc="00000003" with the cached credentials whilst receiving a 401 that would be great. I already had a working workaround (node-red or curl) but the idea was to move everything into OH. Feel free to close this issue if not worth the time "fixing" it.

**Comment by Skinah** [comment]
It may be a bug as the camera/server did state Connection: close and I could be wrong but the keep-alive is needed to reuse a connection and to increment the NC. A work around may be to use HTTP1.0 requests.

**Comment by Skinah** [comment]
I should clarify that it would be a bug in jetty and hence need to be reported at their project and not in openhab if what I believe to be the case is correct. Is the Jetty version up to date does anyone know?

**Comment by J-N-K** [comment]
This is not a bug in Jetty. It's a misconfigured/buggy server.

**Comment by mrjvb** [comment]
> It may be a bug as the camera/server did state Connection: close and I could be wrong but the keep-alive is needed to reuse a connection and to increment the NC. A work around may be to use HTTP1.0 requests.
> 
> When I created the digest code for the ipcamera binding I did try to get the NC feature working, but none of the camera brands supported it. The ipcamera binding can be used to talk to hikvision cameras without issues.

Thx, this is the unfortunate reality for many camera's. Fixing the "server" is not an option unfortunately. I was hoping the http binding (perhaps smarthomej version) could be amended to bypass jetty and obtain new credentials in case of a closed connection. No big deal, there are workarounds like the ipcamera binding.

**Comment by Skinah** [comment]
Another work around as I suggested would be to implement a way to force http1.0 to be used when talking to the server, if that is not already possible as 1.0 can not reuse a connection with NC.

The ipcamera binding uses netty to talk to the cameras and hence I can implement many work arounds that cameras seem to require. It then uses jetty for serving out the normal openhab ports.

Dahua also seems to be in a similar situation based on one camera I tested years ago now, only they state keep-alive and did not allow NC increments when I tried to get it running with netty and my own coding.
